### PR TITLE
Pdm support

### PR DESCRIPTION
**Pull request by betaboon**
This adds support for pdm-lockfiles.

Thanks to the great support of the maintainer of pdm we got in a [PR](https://github.com/pdm-project/pdm/pull/1203) to change their lockfile to contain urls instead of filenames.

### CODE REVIEW

File: pycross/private/tools/pdm_translator.py
Comment (jvolkman) [inline pycross/private/tools/pdm_translator.py:64]: I think this extras handling might be one of the reasons I started using Poetry instead. But I don't recall completely. If you're curious, the earliest revisions in this repo use PDM, I think.

**Review by jvolkman — COMMENTED** [review]
Two early thoughts:

1) let's either combine the PDM example along side the existing `example`, or make an `examples` directory and move them both under there (`examples/poetry`, `examples/pdm`). For the former option, we could move the `:update_example_lock`, `//deps`, etc. under a sub package, like `//pdm:update_example_lock`, `//pdm/deps/...`. Maybe going with the latter approach - separate directories under `examples/` - is better.

2) I ran into this while running `:update_example_lock`
```
OSError: [Errno 30] Read-only file system: '/home/jvolkman/.cache/pdm'
```
Likely due to sandboxing on Linux. We'd need to disable this PDM caching behavior somehow.

File: pycross/private/tools/pdm_translator.py
Comment (betaboon) [reply pycross/private/tools/pdm_translator.py:64]: i agree. I'll change the comments accordingly to document that decision then

**Comment by betaboon** [comment]
>     1. let's either combine the PDM example along side the existing `example`, or make an `examples` directory and move them both under there (`examples/poetry`, `examples/pdm`). For the former option, we could move the `:update_example_lock`, `//deps`, etc. under a sub package, like `//pdm:update_example_lock`, `//pdm/deps/...`. Maybe going with the latter approach - separate directories under `examples/` - is better.

i think the `examples/poetry`+`examples/pdm` approach seems more clean.
question is: do we want to duplicate all the things (eg `deps`, `tools`, `wheels`) to make the examples self-contained (which i think is easier to follow) or do we want to make it shared between the examples? 

>     2. I ran into this while running `:update_example_lock`
> ```
> OSError: [Errno 30] Read-only file system: '/home/jvolkman/.cache/pdm'
> ```
> 
> Likely due to sandboxing on Linux. We'd need to disable this PDM caching behavior somehow.

i can reproduce that issue when running `bazel run :update_example_lock --spawn_strategy=worker,linux-sandbox`.
looking at the code of pdm it seems very hard to deactivate.
i see three possible solutions:
- configure the cache to be in a writable directory, eg `/tmp` (configuration seems to be possible in pdm, but i dunno if writing to `/tmp` is possible with bazel in a sandbox)
- find a "creative" solution (coerce pdm to have the cache in memory, while making it believe it writes on fs)
- send another PR (which seems like a lot of work)

**Review by jvolkman — COMMENTED** [review]
The temp directory solution for PDM's cache seems fine.

We will need to add support for URLs to `lock_model.py` and `bzl_lock_generator.py`. I'm thinking that `PackageFile` can have `urls: Optional[List[str]]`, and if populated we'll create an `http_file` rule rather than `pypi_file`. 

You mentioned at some point that you also have git repository dependencies. Does the change to `pdm` handle those as well? I.e., will there be `git://` urls or something?

**Comment by betaboon** [comment]
> The temp directory solution for PDM's cache seems fine.

I'll leave it like that for now then.

> We will need to add support for URLs to `lock_model.py` and `bzl_lock_generator.py`. I'm thinking that `PackageFile` can have `urls: Optional[List[str]]`, and if populated we'll create an `http_file` rule rather than `pypi_file`.

i was planning to leave that to a new PR, to wait until our changes to PDM are actually available in a release.
but you don't say: pdm `v2.0.0b2` got release 5 hours ago :D
so I'll try to implement your suggestion.

> You mentioned at some point that you also have git repository dependencies. Does the change to `pdm` handle those as well? I.e., will there be `git://` urls or something?

this PR does not yet address git-urls. will try to take of it in the scope of this PR tho

**Comment by jvolkman** [comment]
Feel free to leave things for later, too.

**Comment by betaboon** [comment]
so thanks to your commits for supporting the urls properly this now seems to work.

support for urls in pdm-lockfiles based on the `2.0.0b2` release is now supported here.

so imho this is good to go

**Comment by betaboon** [comment]
having used this for a couple of days, this seems good to go

**Review by jvolkman — APPROVED** [review]
LGTM. Thanks!

**Comment by betaboon** [comment]
as mentioned in slack i just force-pushed to the PR-branch. 

things i changed:
- update to pdm 2.1.1
- rename the args as mentioned above
- add support for groups (defaults to only picking dependencies from the default-group, so as before)

**Pull request merged**
